nemo_aligner/algorithms/self_rewarding.py

```diff
@@ -0,0 +1,1559 @@
+                    del logprobs, chosen_logps, reject_logps, new_batch
+
+                buffer.clear()
+
+                # print(f"*** Rank [ {torch.distributed.get_rank()} ] Iteration [ {self.iteration} ] Step [ {self.step} ] META_BATCH_PENDING [ {len(meta_buffer_pending)} ] META_BATCH_ROLLOUT [ {sum([len(x[-1]) for x in meta_buffer_pending])} ] META_BATCH_DONE [ {len(meta_buffer_done)} ]")
+                # print(f"*** Rank [ {torch.distributed.get_rank()} ] Iteration [ {self.iteration} ] Step [ {self.step} ] META_CNTR  {cnt_tracker}  META_CNTR_PCNT  {cnt_tracker / sum(cnt_tracker).clip(min=1.0)}")
+                if done:
+                    meta_buffer_pending.clear()
+                    # meta_buffer_done.clear()
+                if (
+                    self.use_meta_judge
+                    and (not done)
+                    and (rollout_len := sum([len(x[-1]) for x in meta_buffer_pending]))
+                    >= self.rollout_micro_batch_size
+                ):
+                    num_rollouts = rollout_len // self.rollout_micro_batch_size
+                    meta_buffer_unroll_grp = [(idx, y) for idx, x in enumerate(meta_buffer_pending) for y in x[-1]]
+                    for _ in range(num_rollouts):
+                        meta_buffer_unroll = [
+                            meta_buffer_unroll_grp.pop(0) for _ in range(self.rollout_micro_batch_size)
+                        ]
+                        meta_reward_scores = self.get_rewards_meta([x[-1] for x in meta_buffer_unroll])
+                        meta_pairs = []
+                        reroll = [
+                            (grp, [y[-1] for y in x])
+                            for grp, x in itertools.groupby(meta_buffer_unroll, lambda kk: kk[0])
+                        ]
+                        for tup in reroll:
+                            N = len(tup[-1])
+                            bad_meta_sample = False
+                            # list of tuples of (t,s,e,end), one tuple per self.num_evals_to_average
+                            # we need to find a chosen and reject index in this list
+                            reward_tokens_raw = meta_buffer_pending[tup[0]][0]
+                            p = len(reward_tokens_raw)
+
+                            elo_scores = self.get_elo_scores(p, N, meta_reward_scores)
+                            if len(np.unique(elo_scores)) < p:
+                                bad_meta_sample = True
+                                # print("BAD_META_SAMPLE_1")
+
+                            meta_chosen_idx = np.argmax(elo_scores)
+                            meta_reject_idx = np.argmin(elo_scores)
+
+                            chosen_prompt_len = reward_tokens_raw[meta_chosen_idx][1]
+                            chosen_gen_len = reward_tokens_raw[meta_chosen_idx][2]
+                            chosen_tokens = reward_tokens_raw[meta_chosen_idx][0][:chosen_gen_len]
+                            # chosen_score = 0.
```

Comment:
To be removed (same with `reject_score = 0.` below) -- otherwise it looks a bit like a (confusing) comment